### PR TITLE
feat: dag-jose

### PR DESCRIPTION
**Pull request by oed**
First take of using DagJOSE as the secure object envelope.

### CODE REVIEW

File: spec/spec.md [hunk 1]
```diff
@@ -163,11 +163,28 @@ top-level format, as defined below:
...
-- Try to minimize need for re-encryption when permissions change
-:::
+The identity hub data structure uses DagJOSE as the security wrapper for both signatures and encryption. JOSE is very flexible when it comes to the specific algorithm to use. Furthermore it also supports multiple signatories and multiple recipients when dealing with signed and encrypted data respectively.
+
+#### Signature envelope
+Signed objects are encoded using DagJWS, which requires the payload to be encoded as a CID. The payload itself is stored separately using DagCBOR, or some other IPLD codec. The JWS protected header should include a `kid` property with the value being the DIDUrl of the key that signed the JWS.
+
+#### Encryption envelope
+Encrypted objects are encoded using DagJWE. In order to properly encrypt data with DagJWE the data needs to bw encoded as an *inline CID* (see details below). JWEs can be used for both symmetric and asymmetric encryption, at the wrapper layer we don't make any prescriptions about this. JWEs can also use various different ciphers, we reccomend the use of `xchacha20poly1305` as it is one of the most performat ciphers and there are multiple implementations.
```
Comment (TallTed) [inline spec/spec.md:172]:
```suggestion
Encrypted objects are encoded using DagJWE. In order to properly encrypt data with DagJWE the data needs to bw encoded as an *inline CID* (see details below). JWEs can be used for both symmetric and asymmetric encryption, at the wrapper layer we don't make any prescriptions about this. JWEs can also use various different ciphers, we recommend the use of `xchacha20poly1305`, as it is one of the most performant ciphers, and there are multiple implementations.
```

File: spec/spec.md [hunk 2]
```diff
@@ -163,11 +163,28 @@ top-level format, as defined below:
...
+
+##### Encoding an *inline CID*
+In order to encode an object as an *inline CID* the following algorithm is used:
+
+1. Encode the object using some IPLD codec (e.g. DagCBOR)
```
Comment (TallTed) [inline spec/spec.md:177]:
```suggestion
1. Encode the object using some IPLD codec (e.g., DagCBOR)
```

File: spec/spec.md [hunk 3]
```diff
@@ -163,11 +163,28 @@ top-level format, as defined below:
...
+2. Encode the IPLD bytes from step 1 as an "identity multihash", which is constructued as:
+  - `<hash multicodec><size><bytes>` where,
+  - `hash multicodec` is `00` for "identity"
+  - `size` is the varint bytelength
+  - `bytes` is the bytes from step 1
+3. Create an *inline CID* using the codec, which is encoded as
+  - `<multibase prefix><multicodec cidv1><multicodec content type><multihash>` where,
+  - `multicodec content type` is the IPLD codec used (e.g. DagCBOR)
+  - `multihas` is the bytes from step 2
```
Comment (TallTed) [inline spec/spec.md:186]:
```suggestion
2. Encode the IPLD bytes from step 1 as an _"identity multihash"_, which is 
constructed as `<hash multicodec><size><bytes>`, where —
   - `hash multicodec` is `00` for "identity"
   - `size` is the varint bytelength
   - `bytes` is the bytes from step 1
3. Create an *inline CID* using the codec, which is encoded as 
`<multibase prefix><multicodec cidv1><multicodec content type><multihash>`, 
where —
  - `multicodec content type` is the IPLD codec used (e.g., `DagCBOR`)
  - `multihas` is the bytes from step 2
```

**Review by TallTed — CHANGES_REQUESTED** [review]
markup and typo corrections

File: spec/spec.md [hunk 4]
```diff
@@ -163,19 +163,28 @@ top-level format, as defined below:
...
```
Comment (csuwildcat) [inline spec/spec.md:186]: Should this be `multihash`? Seems to be missing an `h` at the end

**Review by csuwildcat — APPROVED** [review]

**Comment by csuwildcat** [comment]
@TallTed @oed  are the review comments all concluded/pulled in? If so, I can merge the PR in.

**Comment by TallTed** [comment]
@csuwildcat -- @oed needs to accept my suggestions (or justify their rejection).  Then merge can happen.

File: spec/spec.md (hunk 2 above)
Comment (oed) [reply spec/spec.md:177]: This seems weird.

**Review by oed — COMMENTED** [review]
Thanks for the suggestings @TallTed

File: spec/spec.md (hunk 2 above)
Comment (csuwildcat) [reply spec/spec.md:177]: @oed if it's not a suitable change, you can decline the change, which allows us to merge

Comment (oed) [reply spec/spec.md:177]: All good to merge on my end.

Comment (TallTed) [reply spec/spec.md:177]: See https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/

**Pull request merged**